src/libstd/sys/unix/fs_linux.rs

```diff
@@ -0,0 +1,680 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use cell::RefCell;
+use cmp;
+use fs::{File, Metadata};
+use io::{self, Error, ErrorKind, Read, Write};
+use libc;
+use mem;
+use os::linux::fs::MetadataExt;
+use path::Path;
+use ptr;
+use super::ext::io::AsRawFd;
+use sys::{cvt, cvt_r};
+
+unsafe fn copy_file_range(
+    fd_in: libc::c_int,
+    off_in: *mut libc::loff_t,
+    fd_out: libc::c_int,
+    off_out: *mut libc::loff_t,
+    len: libc::size_t,
+    flags: libc::c_uint,
+) -> libc::c_long {
+    libc::syscall(
+        libc::SYS_copy_file_range,
+        fd_in,
+        off_in,
+        fd_out,
+        off_out,
+        len,
+        flags,
+    )
+}
+
+/// Corresponds to lseek(2) `wence`. This exists in std, but doesn't support sparse-files.
+#[allow(dead_code)]
+enum Wence {
+    Set = libc::SEEK_SET as isize,
+    Cur = libc::SEEK_CUR as isize,
+    End = libc::SEEK_END as isize,
+    Data = libc::SEEK_DATA as isize,
+    Hole = libc::SEEK_HOLE as isize,
+}
+
+#[derive(PartialEq, Debug)]
+enum SeekOff {
+    Offset(u64),
+    EOF
+}
+
+fn lseek(fd: &File, off: i64, wence: Wence) -> io::Result<SeekOff> {
+    let r = unsafe {
+        libc::lseek64(
+            fd.as_raw_fd(),
+            off,
+            wence as libc::c_int
+        )
+    };
+
+    if r == -1 {
+        let err = io::Error::last_os_error();
+        match err.raw_os_error() {
+            Some(errno) if errno == libc::ENXIO => {
+                Ok(SeekOff::EOF)
+            }
+            _ => Err(err.into())
+        }
+
+    } else {
+        Ok(SeekOff::Offset(r as u64))
+    }
+
+}
+
+fn allocate_file(fd: &File, len: u64) -> io::Result<()> {
+    cvt_r(|| unsafe {libc::ftruncate64(fd.as_raw_fd(), len as i64)})?;
+    Ok(())
+}
+
+
+// Wrapper for copy_file_range(2) that defers file offset tracking to
+// the underlying call. See the manpage for details.
+fn copy_bytes_kernel(reader: &File, writer: &File, nbytes: usize) -> io::Result<u64> {
+    unsafe {
+        cvt(copy_file_range(reader.as_raw_fd(),
+                            ptr::null_mut(),
+                            writer.as_raw_fd(),
+                            ptr::null_mut(),
+                            nbytes,
+                            0)
+        )
+    }
+    .map(|v| v as u64)
+}
+
+// Slightly modified version of io::copy() that only copies a set amount of bytes.
+fn copy_bytes_uspace(mut reader: &File, mut writer: &File, nbytes: usize) -> io::Result<u64> {
+    const BLKSIZE: usize = 4 * 1024;  // Assume 4k blocks on disk.
```

Comment:
This is a reasonable default for `io::copy` that handles generic `Read` and `Write` types but for `fs::copy` we know we're doing syscalls and can probably do better. E.g. gnu cp has bloated up over the years and [now does 128KiB or stat.blksize](https://github.com/coreutils/coreutils/blob/master/src/ioblksize.h#L73-L78), *whichever is larger*.

This is probably due to ever-growing throughput (thus more syscalls per unit of time) and larger true block sizes, e.g. on flash and raid storage.

Reply:
Oh, and of course it [only does this for large files](https://github.com/coreutils/coreutils/blob/439741053256618eb651e6d43919df29625b8714/src/copy.c#L1257-L1260).

On the other hand the rust std default buffer size was dropped to rust 8k in #32695 due to problems with jemalloc. But that might not be relevant here since jemalloc is not the default anymore and this would only apply to copying large files.